src/blockdata/opcodes.rs

```diff
@@ -1093,5 +1121,75 @@ mod tests {
         roundtrip!(unique, OP_RETURN_255);
         assert_eq!(unique.len(), 256);
     }
+
+    macro_rules! is_in_class {
+        ($op:ident, $cl:expr) => {
+            assert_eq!(All::from(all::$op.into_u8()).classify(), $cl)
+        };
+    }
+
+    #[test]
+    fn classify() {
+        use super::Class as C;
+        is_in_class!(OP_VERIF, C::IllegalOp);
+        is_in_class!(OP_VERNOTIF, C::IllegalOp);
+        is_in_class!(OP_CAT, C::IllegalOp);
+        is_in_class!(OP_SUBSTR, C::IllegalOp);
+        is_in_class!(OP_LEFT, C::IllegalOp);
+        is_in_class!(OP_RIGHT, C::IllegalOp);
+        is_in_class!(OP_INVERT, C::IllegalOp);
+        is_in_class!(OP_AND, C::IllegalOp);
+        is_in_class!(OP_OR, C::IllegalOp);
+        is_in_class!(OP_XOR, C::IllegalOp);
+        is_in_class!(OP_2MUL, C::IllegalOp);
+        is_in_class!(OP_2DIV, C::IllegalOp);
+        is_in_class!(OP_MUL, C::IllegalOp);
+        is_in_class!(OP_DIV, C::IllegalOp);
+        is_in_class!(OP_MOD, C::IllegalOp);
+        is_in_class!(OP_LSHIFT, C::IllegalOp);
+        is_in_class!(OP_RSHIFT, C::IllegalOp);
+
+        is_in_class!(OP_NOP, C::NoOp);
+        is_in_class!(OP_NOP1, C::NoOp);
+        is_in_class!(OP_NOP5, C::NoOp);
+        is_in_class!(OP_NOP10, C::NoOp);
+
+        is_in_class!(OP_RESERVED, C::ReturnOp);
+        is_in_class!(OP_VER, C::ReturnOp);
+        is_in_class!(OP_RETURN, C::ReturnOp);
+        is_in_class!(OP_RESERVED1, C::ReturnOp);
+        is_in_class!(OP_RESERVED2, C::ReturnOp);
+        is_in_class!(OP_RETURN_186, C::ReturnOp);
+        is_in_class!(OP_RETURN_220, C::ReturnOp);
+        is_in_class!(OP_RETURN_255, C::ReturnOp);
+
+        is_in_class!(OP_PUSHNUM_NEG1, C::PushNum(-1));
+
+        is_in_class!(
+            OP_PUSHNUM_1,
```

Comment:
Just a quick question, what happens to 0?

Reply:
not sure i'm following as there is no `OP_PUSHNUM_0`

Reply:
`PUSHBYTES_0` is the same as `PUSHNUM_0` for all intents and purposes, and it will be very confusing if the numeric opcodes include -1 through 16 but not 0.

Reply:
Makes sense. I'm still learning basics of the protocol, but is it the solution to simply add `OP_PUSHNUM_0` here and have it return `Class::PushNum(0)`?

Reply:
No, because it is also the 0-byte push opcode :)

Reply:
Sorry, I'm really not following. What would be the solution here?

Reply:
I guess, we should leave the code as-is since that matches the existing logic. But IMO this is a wart in the library that I'd really rather fix.

On the other hand, there isn't really any good solution. The fact is that not all opcodes can be classified uniquely. (We have a similar problem actually with all the numeric opcodes, which are equivalent to pushdata opcodes but classified differently ... and regardless of that, in many contexts numerics *should* be considered as pushes.)